### ISSUE TITLE
IL ordering bug?

### ISSUE THREAD

**Issue by carlos4242** [avr-rust/rust#51]
[avrio.c.txt](https://github.com/avr-rust/rust/files/989241/avrio.c.txt)
[avrio.ll.txt](https://github.com/avr-rust/rust/files/989242/avrio.ll.txt)
[blink.s.txt](https://github.com/avr-rust/rust/files/989243/blink.s.txt)

look at how the pinMode function from avrio.ll gets converted into assembly in blink.s...

It looks like the last bit of the function in IL...

```
if.then2:                                         ; preds = %if.end
  %.sink30 = select i1 %cmp.i, i8* inttoptr (i16 42 to i8*), i8* inttoptr (i16 36 to i8*)
  %0 = load volatile i8, i8* %.sink30, align 2, !tbaa !1
  %or10 = or i8 %0, %conv3.i
  br label %if.end28

if.else13:                                        ; preds = %if.end
  %neg = xor i8 %conv3.i, -1
  %.sink32 = select i1 %cmp.i, i8* inttoptr (i16 42 to i8*), i8* inttoptr (i16 36 to i8*)
  %1 = load volatile i8, i8* %.sink32, align 2, !tbaa !1
  %and25 = and i8 %1, %neg
  br label %if.end28

if.end28:                                         ; preds = %if.else13, %if.then2
  %and25.sink.off0 = phi i8 [ %and25, %if.else13 ], [ %or10, %if.then2 ]
  %.sink32.sink = phi i8* [ %.sink32, %if.else13 ], [ %.sink30, %if.then2 ]
  store volatile i8 %and25.sink.off0, i8* %.sink32.sink, align 1, !tbaa !1
  br label %cleanup

cleanup:                                          ; preds = %if.else.i, %pinMask.exit, %if.end28
  ret void
}
```

Gets translated into assembly in the final product like this...

```
0000018c <LBB2_12>:
 18c:	b0 e0       	ldi	r27, 0x00	; 0
 18e:	8c 91       	ld	r24, X
 190:	82 2b       	or	r24, r18
 192:	28 2f       	mov	r18, r24
 194:	2c 93       	st	X, r18

00000196 <LBB2_14>:
 196:	df 91       	pop	r29
 198:	cf 91       	pop	r28
 19a:	08 95       	ret

0000019c <LBB2_15>:
 19c:	aa e2       	ldi	r26, 0x2A	; 42

0000019e <LBB2_16>:
 19e:	b0 e0       	ldi	r27, 0x00	; 0
 1a0:	8c 91       	ld	r24, X
 1a2:	20 95       	com	r18
 1a4:	28 23       	and	r18, r24

000001a6 <exit>:
 1a6:	f8 94       	cli
 1a8:	00 c0       	rjmp	.+0      	; 0x1aa <_exit>

000001aa <_exit>:
 1aa:	f8 94       	cli
```

The issue is that in the AND case, the st	X, r18 instruction seems to be missing.  Worse still, there's no jump back to LBB2_14, meaning that in this case the function won't even complete and the program will just fall through to the "exit" code instead.

**Comment by shepmaster** [comment]
Hello! This is the AVR-Rust repository; could you please show the corresponding Rust code that generates the issue? Otherwise, you are probably better off opening an issue on the [LLVM issue tracker](https://bugs.llvm.org/)

**Comment by gergoerdi** [comment]
Isn't this just a duplicate of https://github.com/avr-rust/rust/issues/49? I have a fix for this in my fork, can you try that?

**Comment by dylanmckay** [comment]
> Hello! This is the AVR-Rust repository; could you please show the corresponding Rust code that generates the issue? Otherwise, you are probably better off opening an issue on the LLVM issue tracker

FWIW I told Carl he should post it here. BugZilla would be fine too though, I'm not worried.

**Comment by shepmaster** [comment]
@dylanmckay sounds good! I wasn't sure if you wanted non-Rust stuff reported here, so I figured I'd take the conservative route.

**Comment by dylanmckay** [comment]
Fixed in #49
